<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>`scPCA`: Sparse contrastive principal component analysis • scPCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="`scPCA`: Sparse contrastive principal component analysis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scPCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scpca_intro.html">`scPCA`: Sparse contrastive principal component analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhilBoileau/scPCA">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>
<code>scPCA</code>: Sparse contrastive principal component analysis</h1>
                        <h4 class="author">Philippe Boileau and Nima Hejazi</h4>
            
            <h4 class="date">2019-10-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/PhilBoileau/scPCA/blob/master/vignettes/scpca_intro.Rmd"><code>vignettes/scpca_intro.Rmd</code></a></small>
      <div class="hidden name"><code>scpca_intro.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## Loading required package: magrittr</code></pre>
<pre><code>## Loading required package: lars</code></pre>
<pre><code>## Loaded lars 1.2</code></pre>
<pre><code>## scPCA v1.1.0: Sparse Contrastive Principal Component Analysis</code></pre>
<p>Data pre-processing and exploratory data analysis and are two important steps in the data science life-cycle. As data sets become larger and the signal weaker, their importance increases. Thus, methods that are capable of extracting the signal from such data sets is badly needed. Often, these steps rely on dimensionality reduction techniques to isolate pertinent information in data. However, many of the most commonly-used methods fail to reduce the dimensions of these large and noisy data sets successfully.</p>
<p>Principal component analysis (PCA) is one such method. Although popular for its interpretable results and ease of implementation, PCA’s performance on high-dimensional often leaves much to be desired. Its results on these large datasets have been found to be unstable, and it is often unable to identify variation that is contextually meaningful.</p>
<p>Fortunately, modifications of PCA have been developed to remedy these issues. Namely, sparse PCA (sPCA) was created to increase the stability of the principal component loadings and variable scores in high dimensions, and constrastive PCA (cPCA) was proposed as a method for capturing relevant information in the high-dimensional data <span class="citation">(Abid et al. 2018)</span>.</p>
<p>Although sPCA and cPCA have proven useful in resolving individual shortcomings of PCA, neither is capable of tackling the issues of stability and relevance simultaneously. The goal of this research project is to determine whether a combination of these methods, dubbed sparse constrastive PCA (scPCA), can accomplish this task.</p>
<hr>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>To install the latest stable release of the <code>scPCA</code> package from Bioconductor, use <a href="https://CRAN.R-project.org/package=BiocManager"><code>BiocManager</code></a>:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"scPCA"</span>)</a></code></pre></div>
<p>Note that development of the <code>scPCA</code> package is done via its GitHub repository. If you wish to contribute to the development of the package or use features that have not yet been introduced into a stable release, <code>scPCA</code> may be installed from GitHub using <a href="https://CRAN.R-project.org/package=remotes"><code>remotes</code></a>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">remotes<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remotes/man/install_github.html">install_github</a></span>(<span class="st">"PhilBoileau/scPCA"</span>)</a></code></pre></div>
<hr>
</div>
<div id="comparing-pca-spca-cpca-and-scpca" class="section level1">
<h1 class="hasAnchor">
<a href="#comparing-pca-spca-cpca-and-scpca" class="anchor"></a>Comparing PCA, SPCA, cPCA and scPCA</h1>
<p>A brief comparion of PCA, SPCA, cPCA and scPCA is provided below. All four methods are applied to a simulated target dataset consisting of 400 observations and 30 continuous variables. Additionally, each observation is classified as belonging to one of four classes. This label is known is known a priori. A background data set comprised of the same number of variables as the target data set.</p>
<p>The target data was simulated as follows:</p>
<ul>
<li>Each of the first 10 variables was drawn from <span class="math inline">\(N(0, 10)\)</span>
</li>
<li>For group 1 and 2, variables 11 through 20 were drawn from <span class="math inline">\(N(0, 1)\)</span>
</li>
<li>For group 3 and 4, variables 11 through 20 were drawn from <span class="math inline">\(N(3, 1)\)</span>
</li>
<li>For group 1 and 3, variables 21 though 30 were drawn from <span class="math inline">\(N(-3, 1)\)</span>
</li>
<li>For group 2 and 4, variables 21 though 30 were drawn from <span class="math inline">\(N(0, 1)\)</span>
</li>
</ul>
<p>The background data was simulated as follows:</p>
<ul>
<li>The first 10 variables were drawn from <span class="math inline">\(N(0, 10)\)</span>
</li>
<li>Variables 11 through 20 were drawn from <span class="math inline">\(N(0, 3)\)</span>
</li>
<li>Variables 21 through 30 were drawn from <span class="math inline">\(N(0, 1)\)</span>
</li>
</ul>
<p>A similar simulation scheme is provided in <span class="citation">Abid et al. (2018)</span>.</p>
<div id="pca" class="section level2">
<h2 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h2>
<p>First, PCA is applied to the target data. As we can see from the figure, PCA is incapable of creating a lower dimensional representation of the target data that captures the variation of interest (i.e. the four groups). In fact, no pair of principal components among the first twelve were able to.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># set seed for reproducibility</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">1742</span>)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="co"># load data</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(toy_df)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6"></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="co"># perform PCA</span></a>
<a class="sourceLine" id="cb10-8" data-line-number="8">pca_sim &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(toy_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>])</a>
<a class="sourceLine" id="cb10-9" data-line-number="9"></a>
<a class="sourceLine" id="cb10-10" data-line-number="10"><span class="co"># plot the 2D rep using first 2 components</span></a>
<a class="sourceLine" id="cb10-11" data-line-number="11">df &lt;-<span class="st"> </span><span class="kw">as_tibble</span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"PC1"</span> =<span class="st"> </span>pca_sim<span class="op">$</span>x[, <span class="dv">1</span>],</a>
<a class="sourceLine" id="cb10-12" data-line-number="12">                     <span class="st">"PC2"</span> =<span class="st"> </span>pca_sim<span class="op">$</span>x[, <span class="dv">2</span>],</a>
<a class="sourceLine" id="cb10-13" data-line-number="13">                     <span class="st">"label"</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(toy_df[, <span class="dv">31</span>])))</a>
<a class="sourceLine" id="cb10-14" data-line-number="14">p_pca &lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> PC1, <span class="dt">y =</span> PC2, <span class="dt">colour =</span> label)) <span class="op">+</span></a>
<a class="sourceLine" id="cb10-15" data-line-number="15"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"PCA on Simulated Data"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb10-16" data-line-number="16"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb10-17" data-line-number="17"><span class="st">  </span><span class="kw">theme_minimal</span>()</a>
<a class="sourceLine" id="cb10-18" data-line-number="18">p_pca</a></code></pre></div>
<p><img src="scpca_intro_files/figure-html/PCA_sim-1.png" width="700"></p>
</div>
<div id="sparse-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#sparse-pca" class="anchor"></a>Sparse PCA</h2>
<p>Much like PCA, the leading components of SPCA – for varying amounts of sparsity – are incapable of splitting the observations into four distinct groups.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># perform sPCA on toy_df for a range of L1 penalty terms</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">penalties &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="dv">10</span>), <span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="dv">1000</span>), <span class="dt">length.out =</span> <span class="dv">6</span>))</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">df_ls &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(penalties, <span class="cf">function</span>(penalty) {</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">  spca_sim_p &lt;-<span class="st"> </span><span class="kw">spca</span>(toy_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>], <span class="dt">K =</span> <span class="dv">2</span>, <span class="dt">para =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(penalty, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">                     <span class="dt">type =</span> <span class="st">"predictor"</span>, <span class="dt">sparse =</span> <span class="st">"penalty"</span>)<span class="op">$</span>loadings</a>
<a class="sourceLine" id="cb11-6" data-line-number="6">  spca_sim_p &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(toy_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>]) <span class="op">%*%</span><span class="st"> </span>spca_sim_p</a>
<a class="sourceLine" id="cb11-7" data-line-number="7">  spca_out &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"SPC1"</span> =<span class="st"> </span>spca_sim_p[, <span class="dv">1</span>],</a>
<a class="sourceLine" id="cb11-8" data-line-number="8">                   <span class="st">"SPC2"</span> =<span class="st"> </span>spca_sim_p[, <span class="dv">2</span>],</a>
<a class="sourceLine" id="cb11-9" data-line-number="9">                   <span class="st">"penalty"</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(penalty, <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(toy_df))),</a>
<a class="sourceLine" id="cb11-10" data-line-number="10">                   <span class="st">"label"</span>  =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(toy_df[, <span class="dv">31</span>])) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb11-11" data-line-number="11"><span class="st">    </span><span class="kw">as_tibble</span>()</a>
<a class="sourceLine" id="cb11-12" data-line-number="12">  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(spca_out)</a>
<a class="sourceLine" id="cb11-13" data-line-number="13">})</a>
<a class="sourceLine" id="cb11-14" data-line-number="14">df &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(df_ls)</a>
<a class="sourceLine" id="cb11-15" data-line-number="15"></a>
<a class="sourceLine" id="cb11-16" data-line-number="16"><span class="co"># plot the results of sPCA</span></a>
<a class="sourceLine" id="cb11-17" data-line-number="17">p_spca &lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> SPC1, <span class="dt">y =</span> SPC2, <span class="dt">colour =</span> label)) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-18" data-line-number="18"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-19" data-line-number="19"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"SPCA on Simulated Data for Varying L1 Penalty Terms"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-20" data-line-number="20"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>penalty, <span class="dt">nrow =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-21" data-line-number="21"><span class="st">  </span><span class="kw">theme_minimal</span>()</a>
<a class="sourceLine" id="cb11-22" data-line-number="22">p_spca</a></code></pre></div>
<p><img src="scpca_intro_files/figure-html/sPCA_sim-1.png" width="700"></p>
</div>
<div id="contrastive-pca-cpca" class="section level2">
<h2 class="hasAnchor">
<a href="#contrastive-pca-cpca" class="anchor"></a>Contrastive PCA (cPCA)</h2>
<p>The first two contrastive principal components of cPCA successfully captured the variation of interest in the data with the help of the background data set. To fit contrastive PCA with the <code>scPCA</code> function of this package, simply select no penalization (by setting argument <code>penalties = 0</code>):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># run cPCA across with a penalty of zero to recover Abid et al.'s cPCA</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">cpca_sim &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scPCA.html">scPCA</a></span>(<span class="dt">target =</span> toy_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>],</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">                  <span class="dt">background =</span> background_df,</a>
<a class="sourceLine" id="cb12-4" data-line-number="4">                  <span class="dt">penalties =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb12-5" data-line-number="5">                  <span class="dt">n_centers =</span> <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb12-6" data-line-number="6"></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="co"># create a dataframe to be plotted</span></a>
<a class="sourceLine" id="cb12-8" data-line-number="8">cpca_df &lt;-<span class="st"> </span>cpca_sim<span class="op">$</span>x <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-9" data-line-number="9"><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-10" data-line-number="10"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">label =</span> toy_df[, <span class="dv">31</span>] <span class="op">%&gt;%</span><span class="st"> </span>as.character)</a>
<a class="sourceLine" id="cb12-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(cpca_df) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cPC1"</span>, <span class="st">"cPC2"</span>, <span class="st">"label"</span>)</a>
<a class="sourceLine" id="cb12-12" data-line-number="12"></a>
<a class="sourceLine" id="cb12-13" data-line-number="13"><span class="co"># plot the results</span></a>
<a class="sourceLine" id="cb12-14" data-line-number="14">p_cpca &lt;-<span class="st"> </span><span class="kw">ggplot</span>(cpca_df, <span class="kw">aes</span>(<span class="dt">x =</span> cPC1, <span class="dt">y =</span> cPC2, <span class="dt">colour =</span> label)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-15" data-line-number="15"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-16" data-line-number="16"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"cPCA of Simulated Data"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-17" data-line-number="17"><span class="st">  </span><span class="kw">theme_minimal</span>()</a>
<a class="sourceLine" id="cb12-18" data-line-number="18">p_cpca</a></code></pre></div>
<p><img src="scpca_intro_files/figure-html/cPCA_sim-1.png" width="700"></p>
</div>
<div id="scpca-sparse-contrastive-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#scpca-sparse-contrastive-pca" class="anchor"></a>scPCA: Sparse Contrastive PCA</h2>
<p>The leading sparse contrastive components were also able to capture the variation of interest, though the clusters corresponding to the class labels are more loose than those of cPCA. Importantly, the first and second scPC contain only two and three non-zero loadings, respectively – a significant improvement over cPCA, whose first and second cPCs each possess 30 non-zero loadings, in terms of interpretability.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co"># run scPCA for using 40 logarithmically seperated contrastive parameter values</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="co"># and possible 20 L1 penalty terms</span></a>
<a class="sourceLine" id="cb13-3" data-line-number="3">scpca_sim &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scPCA.html">scPCA</a></span>(<span class="dt">target =</span> toy_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>],</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">                   <span class="dt">background =</span> background_df,</a>
<a class="sourceLine" id="cb13-5" data-line-number="5">                   <span class="dt">n_centers =</span> <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb13-6" data-line-number="6"></a>
<a class="sourceLine" id="cb13-7" data-line-number="7"><span class="co"># create a dataframe to be plotted</span></a>
<a class="sourceLine" id="cb13-8" data-line-number="8">scpca_df &lt;-<span class="st"> </span>scpca_sim<span class="op">$</span>x <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb13-9" data-line-number="9"><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb13-10" data-line-number="10"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">label =</span> toy_df[, <span class="dv">31</span>] <span class="op">%&gt;%</span><span class="st"> </span>as.character)</a>
<a class="sourceLine" id="cb13-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(scpca_df) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"scPC1"</span>, <span class="st">"scPC2"</span>, <span class="st">"label"</span>)</a>
<a class="sourceLine" id="cb13-12" data-line-number="12"></a>
<a class="sourceLine" id="cb13-13" data-line-number="13"><span class="co"># plot the results</span></a>
<a class="sourceLine" id="cb13-14" data-line-number="14">p_scpca &lt;-<span class="st"> </span><span class="kw">ggplot</span>(scpca_df, <span class="kw">aes</span>(<span class="dt">x =</span> scPC1, <span class="dt">y =</span> scPC2, <span class="dt">colour =</span> label)) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-15" data-line-number="15"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-16" data-line-number="16"><span class="st">  </span><span class="kw">theme_minimal</span>()</a>
<a class="sourceLine" id="cb13-17" data-line-number="17">p_scpca</a></code></pre></div>
<p><img src="scpca_intro_files/figure-html/scPCA_sim-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># create the loadings comparison plot</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2">load_diff_df &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">  cpca_sim<span class="op">$</span>rotation <span class="op">%&gt;%</span><span class="st"> </span>as.data.frame,</a>
<a class="sourceLine" id="cb14-4" data-line-number="4">  scpca_sim<span class="op">$</span>rotation <span class="op">%&gt;%</span><span class="st"> </span>as.data.frame) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="st">  </span><span class="kw">mutate</span>(</a>
<a class="sourceLine" id="cb14-6" data-line-number="6">    <span class="dt">sparse =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"0"</span>, <span class="dv">30</span>), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"1"</span>, <span class="dv">30</span>)),</a>
<a class="sourceLine" id="cb14-7" data-line-number="7">    <span class="dt">coef =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">30</span>, <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb14-8" data-line-number="8">  )</a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(load_diff_df) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"comp1"</span>, <span class="st">"comp2"</span>, <span class="st">"sparse"</span>, <span class="st">"coef"</span>)</a>
<a class="sourceLine" id="cb14-10" data-line-number="10"></a>
<a class="sourceLine" id="cb14-11" data-line-number="11">p1 &lt;-<span class="st"> </span>load_diff_df <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-12" data-line-number="12"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">y =</span> comp1, <span class="dt">x =</span> coef, <span class="dt">fill =</span> sparse)) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-13" data-line-number="13"><span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-14" data-line-number="14"><span class="st">  </span><span class="kw">xlab</span>(<span class="st">""</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-15" data-line-number="15"><span class="st">  </span><span class="kw">ylab</span>(<span class="st">""</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-16" data-line-number="16"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-17" data-line-number="17"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"First Component"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-18" data-line-number="18"><span class="st">  </span><span class="kw">scale_fill_discrete</span>(<span class="dt">name =</span> <span class="st">"Method"</span>, <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cPCA"</span>, <span class="st">"scPCA"</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-19" data-line-number="19"><span class="st">  </span><span class="kw">theme_minimal</span>()</a>
<a class="sourceLine" id="cb14-20" data-line-number="20"></a>
<a class="sourceLine" id="cb14-21" data-line-number="21">p2 &lt;-<span class="st"> </span>load_diff_df <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-22" data-line-number="22"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">y =</span> comp2, <span class="dt">x =</span> coef, <span class="dt">fill =</span> sparse)) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-23" data-line-number="23"><span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-24" data-line-number="24"><span class="st">  </span><span class="kw">xlab</span>(<span class="st">""</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-25" data-line-number="25"><span class="st">  </span><span class="kw">ylab</span>(<span class="st">""</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-26" data-line-number="26"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-27" data-line-number="27"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Second Component"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-28" data-line-number="28"><span class="st">  </span><span class="kw">scale_fill_discrete</span>(<span class="dt">name =</span> <span class="st">"Method"</span>, <span class="dt">labels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cPCA"</span>, <span class="st">"scPCA"</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-29" data-line-number="29"><span class="st">  </span><span class="kw">theme_minimal</span>()</a>
<a class="sourceLine" id="cb14-30" data-line-number="30"></a>
<a class="sourceLine" id="cb14-31" data-line-number="31"><span class="co"># build full plot via ggpubr</span></a>
<a class="sourceLine" id="cb14-32" data-line-number="32"><span class="kw">annotate_figure</span>(</a>
<a class="sourceLine" id="cb14-33" data-line-number="33">  <span class="kw">ggarrange</span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>, <span class="dt">ncol =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb14-34" data-line-number="34">            <span class="dt">common.legend =</span> <span class="ot">TRUE</span>, <span class="dt">legend =</span> <span class="st">"right"</span>),</a>
<a class="sourceLine" id="cb14-35" data-line-number="35">  <span class="dt">top =</span> <span class="st">"Leading Loadings Vectors Comparison"</span>,</a>
<a class="sourceLine" id="cb14-36" data-line-number="36">  <span class="dt">left =</span> <span class="st">"Loading Weights"</span>,</a>
<a class="sourceLine" id="cb14-37" data-line-number="37">  <span class="dt">bottom =</span> <span class="st">"Variable Index"</span></a>
<a class="sourceLine" id="cb14-38" data-line-number="38">)</a></code></pre></div>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_bar).

## Warning: Removed 1 rows containing missing values (geom_bar).

## Warning: Removed 1 rows containing missing values (geom_bar).</code></pre>
<p><img src="scpca_intro_files/figure-html/scPCA_sim-2.png" width="700"></p>
<hr>
</div>
</div>
<div id="bioconductor-integration-via-singlecellexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#bioconductor-integration-via-singlecellexperiment" class="anchor"></a>Bioconductor Integration via <code>SingleCellExperiment</code>
</h1>
<p>We now turn to discussing how the tools in the <code>scPCA</code> package can be used more readily with data structures common in computational biology by examining their integration with the <code>SingleCellExperiment</code> container class. For our example, we will use <code>splatter</code> to simulate a scRNA-seq data set using the Splatter framework <span class="citation">(Zappia, Phipson, and Oshlack 2017)</span>. This method simulates a scRNA-seq count matrix by way of a gamma-Poisson hierarchical model, where the mean expression level of gene <span class="math inline">\(g_i,\; i = 1, \ldots, p\)</span> is sampled from a gamma distribution, and the count <span class="math inline">\(x_{i, j}, \; j = 1, \ldots, n\)</span> of cell <span class="math inline">\(c_j\)</span> is sampled from a Poisson distribution with mean equal to the mean expression level of <span class="math inline">\(g_i\)</span>.</p>
<p>To start, let’s load the required packages and create a simple dataset of 300 cells and 500 genes. The cells are evenly split among three biological groups. The samples in two of these groups possess genes that are highly differentially expressed when compared to those in other groups; they comprise the target data. The genes of the third group of cells are less differentially expressed to the genes in the target data, and so it is considered the background data set. A large batch effect is simulated to confound the biological signal.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(splatter)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb16-3" data-line-number="3"></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="co"># Simulate the three groups of cells. Mask cell heterogeneity with batch effect</span></a>
<a class="sourceLine" id="cb16-5" data-line-number="5">params &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/splatter/man/newParams.html">newSplatParams</a></span>(</a>
<a class="sourceLine" id="cb16-6" data-line-number="6">  <span class="dt">seed =</span> <span class="dv">6757293</span>,</a>
<a class="sourceLine" id="cb16-7" data-line-number="7">  <span class="dt">nGenes =</span> <span class="dv">500</span>,</a>
<a class="sourceLine" id="cb16-8" data-line-number="8">  <span class="dt">batchCells =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">150</span>, <span class="dv">150</span>),</a>
<a class="sourceLine" id="cb16-9" data-line-number="9">  <span class="dt">batch.facLoc =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.05</span>),</a>
<a class="sourceLine" id="cb16-10" data-line-number="10">  <span class="dt">batch.facScale =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.05</span>),</a>
<a class="sourceLine" id="cb16-11" data-line-number="11">  <span class="dt">group.prob =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span><span class="op">/</span><span class="dv">3</span>, <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb16-12" data-line-number="12">  <span class="dt">de.prob =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb16-13" data-line-number="13">  <span class="dt">de.downProb =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb16-14" data-line-number="14">  <span class="dt">de.facLoc =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">0.2</span>, <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb16-15" data-line-number="15">  <span class="dt">de.facScale =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">0.2</span>, <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb16-16" data-line-number="16">)</a>
<a class="sourceLine" id="cb16-17" data-line-number="17">sim_sce &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/splatter/man/splatSimulate.html">splatSimulate</a></span>(params, <span class="dt">method =</span> <span class="st">"groups"</span>)</a></code></pre></div>
<p>To proceed, we log-transform the raw counts and retain only the 250 most variable genes. We then split the simulated data into target and background data sets. Our goal here is to demonstrate a typical assessment of scRNA-seq data (and data from similar assays) using the tools made available in the <code>scPCA</code> package. A standard analysis would follow a workflow largely similar to the one below, though without such a computationally convenient data set.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># rank genes by variance</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">n_genes &lt;-<span class="st"> </span><span class="dv">250</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3">vars &lt;-<span class="st"> </span><span class="kw">assay</span>(sim_sce) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="st">  </span>log1p <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"><span class="st">  </span>rowVars</a>
<a class="sourceLine" id="cb17-6" data-line-number="6"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(vars) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sim_sce)</a>
<a class="sourceLine" id="cb17-7" data-line-number="7">vars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(vars, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb17-8" data-line-number="8"></a>
<a class="sourceLine" id="cb17-9" data-line-number="9"><span class="co"># subset SCE to n_genes with highest variance</span></a>
<a class="sourceLine" id="cb17-10" data-line-number="10">sce_sub &lt;-<span class="st"> </span>sim_sce[<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(vars[<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(n_genes)]),]</a>
<a class="sourceLine" id="cb17-11" data-line-number="11">sce_sub</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 250 300 
## metadata(1): Params
## assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts
## rownames(250): Gene336 Gene362 ... Gene409 Gene444
## rowData names(9): Gene BaseGeneMean ... DEFacGroup2 DEFacGroup3
## colnames(300): Cell1 Cell2 ... Cell299 Cell300
## colData names(4): Cell Batch Group ExpLibSize
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co"># split the subsetted SCE into target and background SCEs</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2">tg_sce &lt;-<span class="st"> </span>sce_sub[, sce_sub<span class="op">$</span>Group <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Group1"</span>, <span class="st">"Group3"</span>)]</a>
<a class="sourceLine" id="cb19-3" data-line-number="3">bg_sce &lt;-<span class="st"> </span>sce_sub[, sce_sub<span class="op">$</span>Group <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Group2"</span>)]</a></code></pre></div>
<p>Note that we limit our analysis to just 250 genes in the interest of time, a typical analysis would generally include a much larger proportion (if not all) of the genes assayed.</p>
<p>Owing to the flexibility of the <code>SingleCellExperiment</code> class, we are able to generate PCA, cPCA, and scPCA representations of the target data, storing these in <code>SingleCellExperiment</code> object using the <code>reducedDims</code> method.</p>
<p>Below, we perform standard PCA on the log-transformed target data, which has been centered and scaled, and perform both cPCA and cPCA using the <code>scPCA</code> function, storing each in a separate object. After applying each of these dimension reduction techniques, we store the resultant objects in a <code>SimpleList</code> that is then appended to the <code>SingleCellExperiment</code> object using the <code>reducedDims</code> accessor. The results are presented in the following figure. cPCA and scPCA successfully remove the batch effect, though PCA is incapable of doing so in two dimensions.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co"># for both cPCA and scPCA, we'll set the penalties and contrasts arguments</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">contrasts &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">0.1</span>), <span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="dv">100</span>), <span class="dt">length.out =</span> <span class="dv">5</span>))</a>
<a class="sourceLine" id="cb20-3" data-line-number="3">penalties &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">0.2</span>, <span class="dv">1</span>, <span class="dt">length.out =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb20-4" data-line-number="4"></a>
<a class="sourceLine" id="cb20-5" data-line-number="5"><span class="co"># first, PCA</span></a>
<a class="sourceLine" id="cb20-6" data-line-number="6">pca_out &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(tg_sce))), <span class="dt">center =</span> <span class="ot">TRUE</span>, <span class="dt">scale. =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb20-7" data-line-number="7"></a>
<a class="sourceLine" id="cb20-8" data-line-number="8"><span class="co"># next, cPCA</span></a>
<a class="sourceLine" id="cb20-9" data-line-number="9">cpca_out &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scPCA.html">scPCA</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(tg_sce))),</a>
<a class="sourceLine" id="cb20-10" data-line-number="10">                  <span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(bg_sce))),</a>
<a class="sourceLine" id="cb20-11" data-line-number="11">                  <span class="dt">n_centers =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb20-12" data-line-number="12">                  <span class="dt">n_eigen =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb20-13" data-line-number="13">                  <span class="dt">contrasts =</span> contrasts,</a>
<a class="sourceLine" id="cb20-14" data-line-number="14">                  <span class="dt">penalties =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb20-15" data-line-number="15">                  <span class="dt">center =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb20-16" data-line-number="16">                  <span class="dt">scale =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb20-17" data-line-number="17"></a>
<a class="sourceLine" id="cb20-18" data-line-number="18"><span class="co"># finally, scPCA</span></a>
<a class="sourceLine" id="cb20-19" data-line-number="19">scpca_out &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scPCA.html">scPCA</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(tg_sce))),</a>
<a class="sourceLine" id="cb20-20" data-line-number="20">                   <span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(bg_sce))),</a>
<a class="sourceLine" id="cb20-21" data-line-number="21">                   <span class="dt">n_centers =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb20-22" data-line-number="22">                   <span class="dt">n_eigen =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb20-23" data-line-number="23">                   <span class="dt">contrasts =</span> contrasts,</a>
<a class="sourceLine" id="cb20-24" data-line-number="24">                   <span class="dt">penalties =</span> penalties,</a>
<a class="sourceLine" id="cb20-25" data-line-number="25">                   <span class="dt">center =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb20-26" data-line-number="26">                   <span class="dt">scale =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb20-27" data-line-number="27"></a>
<a class="sourceLine" id="cb20-28" data-line-number="28"><span class="co"># store new representations in the SingleCellExperiment object</span></a>
<a class="sourceLine" id="cb20-29" data-line-number="29"><span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-reducedDims.html">reducedDims</a></span>(tg_sce) &lt;-<span class="st"> </span><span class="kw">SimpleList</span>(<span class="dt">PCA =</span> pca_out<span class="op">$</span>x[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],</a>
<a class="sourceLine" id="cb20-30" data-line-number="30">                                  <span class="dt">cPCA =</span> cpca_out<span class="op">$</span>x,</a>
<a class="sourceLine" id="cb20-31" data-line-number="31">                                  <span class="dt">scPCA =</span> scpca_out<span class="op">$</span>x)</a>
<a class="sourceLine" id="cb20-32" data-line-number="32">tg_sce</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 250 194 
## metadata(1): Params
## assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts
## rownames(250): Gene336 Gene362 ... Gene409 Gene444
## rowData names(9): Gene BaseGeneMean ... DEFacGroup2 DEFacGroup3
## colnames(194): Cell3 Cell5 ... Cell299 Cell300
## colData names(4): Cell Batch Group ExpLibSize
## reducedDimNames(3): PCA cPCA scPCA
## spikeNames(0):</code></pre>
<p><img src="scpca_intro_files/figure-html/plot_red_dims-1.png" width="1600px"></p>
<p>In the above, we set <code>n_eigen = 2</code> in the calls to the <code>scPCA</code> function that generate the cPCA and scPCA output, recovering the rotated gene-level data for just the first two components of the dimension reduction. While this is done to be explicit (as <code>n_eigen = 2</code> by default), we wish to emphasize that it will often be appropriate to set this to a higher value in order to recover further dimensions generated by these techniques, as such additional information may be useful in exploring further signal in the data at hand. For congruence with cPCA and scPCA, we retain only the first two dimensions generated by PCA in the information stored in the <code>SingleCellExperiment</code> object.</p>
<hr>
</div>
<div id="parallelization" class="section level1">
<h1 class="hasAnchor">
<a href="#parallelization" class="anchor"></a>Parallelization</h1>
<p>Finally, note that the <code>scPCA</code> function has an argument <code>parallel</code> (set to <code>FALSE</code> by default), which facilitates parallelized computation of the various subroutines required in constructing the output of the <code>scPCA</code> function. In a standard analysis of genomic data, use of this parallelization will be crucial, thus, each of the core subroutines of <code>scPCA</code> has an equivalent parallelized variant that makes use of the infrastructure provided by the <a href="https://bioconductor.org/packages/BiocParallel"><code>BiocParallel</code> package</a>. In order to make effective use of this paralleization, one need only set <code>parallel = TRUE</code> in a call to <code>scPCA</code> after having registered a particular parallelization back-end for parallel evaluation as described in the <code>BiocParallel</code> documentation. An example of this form of parallelization follows:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># perform the same operations in parallel using BiocParallel</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BiocParallel)</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">param &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span>()</a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span>(param)</a>
<a class="sourceLine" id="cb22-5" data-line-number="5"></a>
<a class="sourceLine" id="cb22-6" data-line-number="6"><span class="co"># perfom cPCA</span></a>
<a class="sourceLine" id="cb22-7" data-line-number="7">cpca_bp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scPCA.html">scPCA</a></span>(</a>
<a class="sourceLine" id="cb22-8" data-line-number="8">  <span class="dt">target =</span> toy_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>],</a>
<a class="sourceLine" id="cb22-9" data-line-number="9">  <span class="dt">background =</span> background_df,</a>
<a class="sourceLine" id="cb22-10" data-line-number="10">  <span class="dt">contrasts =</span> <span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">0.1</span>), <span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="dv">100</span>), <span class="dt">length.out =</span> <span class="dv">10</span>)),</a>
<a class="sourceLine" id="cb22-11" data-line-number="11">  <span class="dt">penalties =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb22-12" data-line-number="12">  <span class="dt">n_centers =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb22-13" data-line-number="13">  <span class="dt">parallel =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb22-14" data-line-number="14">)</a>
<a class="sourceLine" id="cb22-15" data-line-number="15"></a>
<a class="sourceLine" id="cb22-16" data-line-number="16"><span class="co"># perform scPCA</span></a>
<a class="sourceLine" id="cb22-17" data-line-number="17">scpca_bp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scPCA.html">scPCA</a></span>(</a>
<a class="sourceLine" id="cb22-18" data-line-number="18">  <span class="dt">target =</span> toy_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">30</span>],</a>
<a class="sourceLine" id="cb22-19" data-line-number="19">  <span class="dt">background =</span> background_df,</a>
<a class="sourceLine" id="cb22-20" data-line-number="20">  <span class="dt">contrasts =</span> <span class="kw"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">0.1</span>), <span class="kw"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="dv">100</span>), <span class="dt">length.out =</span> <span class="dv">10</span>)),</a>
<a class="sourceLine" id="cb22-21" data-line-number="21">  <span class="dt">penalties =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">0.1</span>, <span class="dv">1</span>, <span class="dt">length.out =</span> <span class="dv">9</span>),</a>
<a class="sourceLine" id="cb22-22" data-line-number="22">  <span class="dt">n_centers =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb22-23" data-line-number="23">  <span class="dt">parallel =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb22-24" data-line-number="24">)</a></code></pre></div>
<hr>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] splatter_1.8.0              SingleCellExperiment_1.6.0 
##  [3] SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
##  [5] BiocParallel_1.18.1         matrixStats_0.55.0         
##  [7] Biobase_2.44.0              GenomicRanges_1.36.1       
##  [9] GenomeInfoDb_1.20.0         IRanges_2.18.3             
## [11] S4Vectors_0.22.1            BiocGenerics_0.30.0        
## [13] scPCA_1.1.0                 elasticnet_1.1.1           
## [15] lars_1.2                    ggpubr_0.2.3               
## [17] magrittr_1.5                ggplot2_3.2.1              
## [19] dplyr_0.8.3                 BiocStyle_2.12.0           
## 
## loaded via a namespace (and not attached):
##  [1] viridisLite_0.3.0      Rdpack_0.11-0          assertthat_0.2.1      
##  [4] BiocManager_1.30.9     GenomeInfoDbData_1.2.1 yaml_2.2.0            
##  [7] globals_0.12.4         pillar_1.4.2           backports_1.1.5       
## [10] lattice_0.20-38        glue_1.3.1             digest_0.6.22         
## [13] checkmate_1.9.4        ggsignif_0.6.0         XVector_0.24.0        
## [16] colorspace_1.4-1       cowplot_1.0.0          htmltools_0.4.0       
## [19] Matrix_1.2-17          pkgconfig_2.0.3        bibtex_0.4.2          
## [22] listenv_0.7.0          bookdown_0.14          zlibbioc_1.30.0       
## [25] purrr_0.3.3            scales_1.0.0           tibble_2.1.3          
## [28] withr_2.1.2            lazyeval_0.2.2         crayon_1.3.4          
## [31] memoise_1.1.0          evaluate_0.14          fs_1.3.1              
## [34] future_1.14.0          MASS_7.3-51.4          tools_3.6.1           
## [37] data.table_1.12.6      gbRd_0.4-11            stringr_1.4.0         
## [40] origami_1.0.2          kernlab_0.9-27         locfit_1.5-9.1        
## [43] munsell_0.5.0          cluster_2.1.0          compiler_3.6.1        
## [46] pkgdown_1.4.1          rlang_0.4.1            grid_3.6.1            
## [49] RCurl_1.95-4.12        bitops_1.0-6           labeling_0.3          
## [52] rmarkdown_1.16         gtable_0.3.0           codetools_0.2-16      
## [55] abind_1.4-5            R6_2.4.0               gridExtra_2.3         
## [58] knitr_1.25             future.apply_1.3.0     rprojroot_1.3-2       
## [61] desc_1.2.0             stringi_1.4.3          Rcpp_1.0.2            
## [64] tidyselect_0.2.5       xfun_0.10</code></pre>
<hr>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-abid2018exploring">
<p>Abid, Abubakar, Martin J Zhang, Vivek K Bagaria, and James Zou. 2018. “Exploring Patterns Enriched in a Dataset with Contrastive Principal Component Analysis.” <em>Nature Communications</em> 9 (1): 2134.</p>
</div>
<div id="ref-zappia2017splatter">
<p>Zappia, Luke, Belinda Phipson, and Alicia Oshlack. 2017. “Splatter: Simulation of Single-Cell Rna Sequencing Data.” <em>Genome Biology</em> 18 (1). BioMed Central: 174.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li>
<a href="#comparing-pca-spca-cpca-and-scpca">Comparing PCA, SPCA, cPCA and scPCA</a><ul class="nav nav-pills nav-stacked">
<li><a href="#pca">PCA</a></li>
      <li><a href="#sparse-pca">Sparse PCA</a></li>
      <li><a href="#contrastive-pca-cpca">Contrastive PCA (cPCA)</a></li>
      <li><a href="#scpca-sparse-contrastive-pca">scPCA: Sparse Contrastive PCA</a></li>
      </ul>
</li>
      <li><a href="#bioconductor-integration-via-singlecellexperiment">Bioconductor Integration via <code>SingleCellExperiment</code></a></li>
      <li><a href="#parallelization">Parallelization</a></li>
      <li><a href="#session-information">Session Information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Philippe Boileau, Nima Hejazi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
