<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sparse contrastive principal components analysis with <code>scPCA</code> • scPCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Sparse contrastive principal components analysis with &lt;code&gt;scPCA&lt;/code&gt;">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scPCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scpca_intro.html">Sparse contrastive principal components analysis with `scPCA`</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhilBoileau/scPCA">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Sparse contrastive principal components analysis with <code>scPCA</code>
</h1>
                        <h4 class="author">Philippe Boileau and <a href="https://nimahejazi.org">Nima Hejazi</a>
</h4>
            
            <h4 class="date">2019-08-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/PhilBoileau/scPCA/blob/master/vignettes/scpca_intro.Rmd"><code>vignettes/scpca_intro.Rmd</code></a></small>
      <div class="hidden name"><code>scpca_intro.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## Loading required package: magrittr</code></pre>
<pre><code>## Loading required package: lars</code></pre>
<pre><code>## Loaded lars 1.2</code></pre>
<p>Data pre-processing and exploratory data analysis and are two important steps in the data science life-cycle. As data sets become larger and the signal weaker, their importance increases. Thus, methods that are capable of extracting the signal from such data sets is badly needed. Often, these steps rely on dimensionality reduction techniques to isolate pertinent information in data. However, many of the most commonly-used methods fail to reduce the dimensions of these large and noisy data sets successfully.</p>
<p>Principal component analysis (PCA) is one such method. Although popular for its interpretable results and ease of implementation, PCA’s performance on high-dimensional often leaves much to be desired. Its results on these large datasets have been found to be unstable, and it is often unable to identify variation that is contextually meaningful.</p>
<p>Fortunately, modifications of PCA have been developed to remedy these issues. Namely, sparse PCA (sPCA) was created to increase the stability of the principal component loadings and variable scores in high dimensions, and constrastive PCA (cPCA) was proposed as a method for capturing relevant information in the high-dimensional data <span class="citation">(Abid et al. 2018)</span>.</p>
<p>Although sPCA and cPCA have proven useful in resolving individual shortcomings of PCA, neither is capable of tackling the issues of stability and relevance simultaneously. The goal of this research project is to determine whether a combination of these methods, dubbed sparse constrastive PCA (scPCA), can accomplish this task.</p>
<hr>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To install the latest stable release of the <code>scPCA</code> package from Bioconductor, use <a href="https://CRAN.R-project.org/package=BiocManager"><code>BiocManager</code></a>:</p>
<p>Note that development of the <code>scPCA</code> package is done via its GitHub repository. If you wish to contribute to the development of the package or use features that have not yet been introduced into a stable release, <code>scPCA</code> may be installed from GitHub using <a href="https://CRAN.R-project.org/package=remotes"><code>remotes</code></a>:</p>
<hr>
</div>
<div id="comparing-pca-spca-cpca-and-scpca" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-pca-spca-cpca-and-scpca" class="anchor"></a>Comparing PCA, SPCA, cPCA and scPCA</h2>
<p>A brief comparion of PCA, SPCA, cPCA and scPCA is provided below. All four methods are applied to a simulated target dataset consisting of 400 observations and 30 continuous variables. Additionally, each observation is classified as belonging to one of four classes. This label is known is known a priori. A background data set comprised of the same number of variables as the target data set.</p>
<p>The target data was simulated as follows:</p>
<ul>
<li>Each of the first 10 variables was drawn from <span class="math inline">\(N(0, 10)\)</span>
</li>
<li>For group 1 and 2, variables 11 through 20 were drawn from <span class="math inline">\(N(0, 1)\)</span>
</li>
<li>For group 3 and 4, variables 11 through 20 were drawn from <span class="math inline">\(N(3, 1)\)</span>
</li>
<li>For group 1 and 3, variables 21 though 30 were drawn from <span class="math inline">\(N(-3, 1)\)</span>
</li>
<li>For group 2 and 4, variables 21 though 30 were drawn from <span class="math inline">\(N(0, 1)\)</span>
</li>
</ul>
<p>The background data was simulated as follows:</p>
<ul>
<li>The first 10 variables were drawn from <span class="math inline">\(N(0, 10)\)</span>
</li>
<li>Variables 11 through 20 were drawn from <span class="math inline">\(N(0, 3)\)</span>
</li>
<li>Variables 21 through 30 were drawn from <span class="math inline">\(N(0, 1)\)</span>
</li>
</ul>
<p>A similar simulation scheme is provided in <span class="citation">Abid et al. (2018)</span>.</p>
<div id="pca" class="section level3">
<h3 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h3>
<p>First, PCA is applied to the target data. As we can see from the figure, PCA is incapable of creating a lower dimensional representation of the target data that captures the variation of interest (i.e. the four groups). In fact, no pair of principal components among the first twelve were able to.</p>
<p><img src="scpca_intro_files/figure-html/PCA_sim-1.png" width="700"></p>
</div>
<div id="sparse-pca" class="section level3">
<h3 class="hasAnchor">
<a href="#sparse-pca" class="anchor"></a>Sparse PCA</h3>
<p>Much like PCA, the leading components of SPCA – for varying amounts of sparsity – are incapable of splitting the observations into four distinct groups.</p>
<p><img src="scpca_intro_files/figure-html/sPCA_sim-1.png" width="700"></p>
</div>
<div id="contrastive-pca-cpca" class="section level3">
<h3 class="hasAnchor">
<a href="#contrastive-pca-cpca" class="anchor"></a>Contrastive PCA (cPCA)</h3>
<p>The first two contrastive principal components of cPCA successfully captured the variation of interest in the data with the help of the background data set. To fit contrastive PCA with the <code>scPCA</code> function of this package, simply select no penalization (by setting argument <code>penalties = 0</code>):</p>
<p><img src="scpca_intro_files/figure-html/cPCA_sim-1.png" width="700"></p>
</div>
<div id="scpca-sparse-contrastive-pca" class="section level3">
<h3 class="hasAnchor">
<a href="#scpca-sparse-contrastive-pca" class="anchor"></a>scPCA: Sparse Contrastive PCA</h3>
<p>The leading sparse contrastive components were also able to capture the variation of interest, though the clusters corresponding to the class labels are more loose than those of cPCA. Importantly, the first and second scPC contain only two and three non-zero loadings, respectively. This is a significant improvement over cPCA, whose first and second cPCs each possess 30 non-zero loadings, in terms of interpretability.</p>
<p><img src="scpca_intro_files/figure-html/scPCA_sim-1.png" width="700"></p>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_bar).

## Warning: Removed 1 rows containing missing values (geom_bar).

## Warning: Removed 1 rows containing missing values (geom_bar).</code></pre>
<p><img src="scpca_intro_files/figure-html/scPCA_sim-2.png" width="700"></p>
<hr>
</div>
</div>
<div id="bioconductor-integration-via-singlecellexperiment" class="section level2">
<h2 class="hasAnchor">
<a href="#bioconductor-integration-via-singlecellexperiment" class="anchor"></a>Bioconductor Integration via <code>SingleCellExperiment</code>
</h2>
<p>We now turn to discussing how the tools in the <code>scPCA</code> package can be used more readily with data structures common in computational biology by examining their integration with the <code>SingleCellExperiment</code> container class. For our example, we will use the <code>allen</code> data from the <code>scRNAseq</code> package, just as was done to demonstrate use of the <code>SingleCellExperiment</code> container in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_creating_singlecellexperiment_instances">vignette from that package</a>.</p>
<p>To start, let’s load the required packages and create a simple <code>SingleCellExperiment</code> instance, including information about spike-ins:</p>
<p>While interesting, the <code>allen</code> data set is rather large. To proceed, we use only a small subset of the data, limited to the 100 genes with the highest variance, after log-transforming the raw counts. Our goal here is to demonstrate a typical assessment of single-cell RNA-seq data (and data from similar assays) using the tools made available in the <code>scPCA</code> package, a standard analysis would follow a workflow largely similar to the above, though without such convenient reductions in the complexity of the available data.</p>
<pre><code>## class: SingleCellExperiment 
## dim: 100 379 
## metadata(2): SuppInfo which_qc
## assays(4): tophat_counts cufflinks_fpkm rsem_counts rsem_tpm
## rownames(100): Lamp5 Fam19a1 ... Rnf2 Zfp35
## rowData names(0):
## colnames(379): SRR2140028 SRR2140022 ... SRR2139341 SRR2139336
## colData names(22): NREADS NALIGNED ... Animal.ID
##   passes_qc_checks_s
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<p>Note that we limit our analysis to just 100 genes in the interest of time, a typical analysis would generally include a much larger proportion (if not all) of the genes assayed.</p>
<p>In order to use either contrastive PCA or its sparse variant, it’s necessary to have a background data set. Absent such a version of the <code>allen</code> data, we use the 100 genes with the lowest variance as the background:</p>
<p>Owing to the flexibility of the <code>SingleCellExperiment</code> class, we are able to generate PCA, cPCA, and scPCA representations of the reduced <code>allen</code> data, storing these in <code>SingleCellExperiment</code> object using the <code>reducedDims</code> method.</p>
<p>Below, we perform standard PCA on the experimental subset from the <code>allen</code> data and perform both cPCA and scPCA using the <code>scPCA</code> function, storing each in a separate object. After applying each of these dimension reduction techniques, we store the resultant objects in a <code>SimpleList</code> that is then appended to the <code>SingleCellExperiment</code> object using the <code>reducedDims</code> accessor.</p>
<pre><code>## class: SingleCellExperiment 
## dim: 100 379 
## metadata(2): SuppInfo which_qc
## assays(4): tophat_counts cufflinks_fpkm rsem_counts rsem_tpm
## rownames(100): Lamp5 Fam19a1 ... Rnf2 Zfp35
## rowData names(0):
## colnames(379): SRR2140028 SRR2140022 ... SRR2139341 SRR2139336
## colData names(22): NREADS NALIGNED ... Animal.ID
##   passes_qc_checks_s
## reducedDimNames(3): PCA cPCA scPCA
## spikeNames(0):</code></pre>
<p>In the above, we set <code>n_eigen = 2</code> in the calls to the <code>scPCA</code> function that generate the cPCA and scPCA output, recovering the rotated gene-level data for just the first two components of the dimension reduction. While this is done to be explicit (as <code>n_eigen = 2</code> by default), we wish to emphasize that it will often be appropriate to set this to a higher value in order to recover further dimensions generated by these techniques, as such additional information may be useful in exploring further signal in the data at hand. For congruence with cPCA and scPCA, we retain only the first two dimensions generated by PCA in the information stored in the <code>SingleCellExperiment</code> object.</p>
<hr>
</div>
<div id="parallelization" class="section level2">
<h2 class="hasAnchor">
<a href="#parallelization" class="anchor"></a>Parallelization</h2>
<p>Finally, note that the <code>scPCA</code> function has an argument <code>parallel</code> (set to <code>FALSE</code> by default), which facilitates parallelized computation of the various subroutines required in constructing the output of the <code>scPCA</code> function. In a standard analysis of genomic data, use of this parallelization will be crucial, thus, each of the core subroutines of <code>scPCA</code> has an equivalent parallelized variant that makes use of the infrastructure provided by the <a href="https://bioconductor.org/packages/BiocParallel"><code>BiocParallel</code> package</a>. In order to make effective use of this paralleization, one need only set <code>parallel = TRUE</code> in a call to <code>scPCA</code> after having registered a particular parallelization back-end for parallel evaluation as described in the <code>BiocParallel</code> documentation. An example of this form of parallelization follows:</p>
<hr>
</div>
<div id="session-information" class="section level2">
<h2 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h2>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] SingleCellExperiment_1.6.0  scRNAseq_1.10.0            
##  [3] SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
##  [5] BiocParallel_1.18.1         matrixStats_0.54.0         
##  [7] Biobase_2.44.0              GenomicRanges_1.36.0       
##  [9] GenomeInfoDb_1.20.0         IRanges_2.18.2             
## [11] S4Vectors_0.22.0            BiocGenerics_0.30.0        
## [13] scPCA_0.0.1                 elasticnet_1.1.1           
## [15] lars_1.2                    ggpubr_0.2.2               
## [17] magrittr_1.5                ggplot2_3.2.1              
## [19] dplyr_0.8.3                 BiocStyle_2.12.0           
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.2             lattice_0.20-38        assertthat_0.2.1      
##  [4] rprojroot_1.3-2        digest_0.6.20          R6_2.4.0              
##  [7] backports_1.1.4        evaluate_0.14          pillar_1.4.2          
## [10] Rdpack_0.11-0          zlibbioc_1.30.0        rlang_0.4.0           
## [13] lazyeval_0.2.2         kernlab_0.9-27         Matrix_1.2-17         
## [16] rmarkdown_1.15         pkgdown_1.3.0.9000     desc_1.2.0            
## [19] labeling_0.3           stringr_1.4.0          RCurl_1.95-4.12       
## [22] munsell_0.5.0          compiler_3.6.1         xfun_0.9              
## [25] pkgconfig_2.0.2        htmltools_0.3.6        tidyselect_0.2.5      
## [28] tibble_2.1.3           gridExtra_2.3          GenomeInfoDbData_1.2.1
## [31] roxygen2_6.1.1         bookdown_0.13          crayon_1.3.4          
## [34] withr_2.1.2            MASS_7.3-51.4          bitops_1.0-6          
## [37] commonmark_1.7         grid_3.6.1             gtable_0.3.0          
## [40] scales_1.0.0           bibtex_0.4.2           stringi_1.4.3         
## [43] XVector_0.24.0         ggsignif_0.6.0         fs_1.3.1              
## [46] xml2_1.2.2             cowplot_1.0.0          tools_3.6.1           
## [49] glue_1.3.1             purrr_0.3.2            yaml_2.2.0            
## [52] colorspace_1.4-1       cluster_2.1.0          BiocManager_1.30.4    
## [55] gbRd_0.4-11            memoise_1.1.0          knitr_1.24</code></pre>
<hr>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-abid2018exploring">
<p>Abid, Abubakar, Martin J Zhang, Vivek K Bagaria, and James Zou. 2018. “Exploring Patterns Enriched in a Dataset with Contrastive Principal Component Analysis.” <em>Nature Communications</em> 9 (1): 2134.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#comparing-pca-spca-cpca-and-scpca">Comparing PCA, SPCA, cPCA and scPCA</a></li>
      <li><a href="#bioconductor-integration-via-singlecellexperiment">Bioconductor Integration via <code>SingleCellExperiment</code></a></li>
      <li><a href="#parallelization">Parallelization</a></li>
      <li><a href="#session-information">Session Information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Philippe Boileau, Nima Hejazi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
